interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/typer/TypeMethodReference.scala

```diff
@@ -56,41 +69,23 @@ class TypeMethodReference(methodName: String, invocationTarget: TypingResult, ca
     clazzMethods match {
       //Static method can be invoked - we cannot find them ATM
       case Nil if isClass => Left(None)
-      case Nil => Left(Some(s"Unknown method '$methodName' in $displayableType"))
-      case methodInfoes => Right(methodInfoes)
+      case Nil => Left(Some(UnknownMethodError(methodName, displayableType)))
+      case methodInfos => Right(methodInfos)
     }
   }
 
-  private def validateMethodParameterTypes(methodInfoes: List[MethodInfo]): Either[Option[String], List[TypingResult]] = {
-    val returnTypesForMatchingMethods = methodInfoes.flatMap { m =>
-      lazy val allMatching = m.parameters.map(_.refClazz).zip(calledParams).forall {
-        case (declaredType, passedType) => passedType.canBeSubclassOf(declaredType)
-      }
-      if (m.parameters.size == calledParams.size && allMatching) Some(m.refClazz) else checkForVarArgs(m)
-    }
-    returnTypesForMatchingMethods match {
-      case Nil =>
-        def toSignature(params: List[TypingResult]) = params.map(_.display).mkString(s"$methodName(", ", ", ")")
-        val methodVariances = methodInfoes.map(_.parameters.map(_.refClazz))
-        Left(Some(s"Mismatch parameter types. Found: ${toSignature(calledParams)}. Required: ${methodVariances.map(toSignature).mkString(" or ")}"))
-      case nonEmpty =>
-        Right(nonEmpty)
+  private def validateMethodParameterTypes(methodInfos: List[MethodInfo]): Either[Option[SpelParseError], List[TypingResult]] = {
+    val returnTypesForMatchingMethods = methodInfos.map(_.apply(calledParams))
+    val combinedReturnTypes = returnTypesForMatchingMethods.map(x => x.map(List(_))).reduce((x, y) => (x, y) match {
```

Comment:
I think you can use Ior here ;)

Reply:
I tried to change it to Ior, but it didn't work as intended and I haven't had time to make it work, so I kept this version.

Reply:
:+1: